### PR TITLE
feat: Implement Rate Limiting 

### PR DESCRIPTION
**Pull request by istarkov**
## Description

This PR integrates a rate-limiting mechanism into the app using Vercel's edge-layer middleware in combination with the Hono framework. This middleware is written in JS with full JSDoc TypeScript typings.

### Architecture & Strategy
Vercel Middleware: Serves as the front-line defense, running at the edge for optimal performance.
Hono Framework: For granular control, we utilize Hono middlewares for applying or skipping rate limits.
Additionally, non-expensive routes benefit from an ephemeral cache to further optimize performance

### Current Limitations
Due to a dependency on CommonJS, we cannot use TypeScript fully. Consequently, session configurations are not yet shared across multiple files. All utility functions are present in a single file for the time being.

### Logic Breakdown
- Assets: Rate limiting skipped.
- Service Calls: Rate limiting skipped.
- /rest/ai/* Endpoint: Rate limit of 20 requests per 5 minutes, identified by userId or IP.
- Unauthorized Users: Rate limit of 5 requests per 5 seconds, based solely on IP.
- Authorized Users: Rate limit of 20 requests per 5 seconds, based on userId.

### Next Steps
Plan to migrate to ESM for enabling TypeScript support for better maintainability.
Use multiple regional KV

### Debug endpoint.
[Rate limit debug endpoint](https://webstudio-builder-kscynssns-webstudio-is.vercel.app/rate-limit-debug)

## Steps for reproduction

1. click button
2. expect xyz

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

File: pnpm-lock.yaml [hunk 1]
```diff
@@ -11788,6 +11915,15 @@ packages:
     resolution: {integrity: sha512-WhB9zCku7EGTj/HQQRz5aUQEUeoQZH2bWcltRErOpymJ4boYE6wL9Tbr23krRPSZ+C5zqNSrSw+Cc7sZZ4b7vg==}
     engines: {node: '>=0.10.0'}
 
+  /isomorphic-fetch@3.0.0:
```
Comment (TrySound) [inline pnpm-lock.yaml:11918]: Not cool. Dead project.

File: apps/builder/package.json [hunk 2]
```diff
@@ -31,12 +34,16 @@
     "@radix-ui/react-tooltip": "^1.0.6",
     "@react-aria/interactions": "^3.17.0",
     "@react-aria/utils": "^3.19.0",
+    "@remix-run/cloudflare": "^2.0.1",
```
Comment (TrySound) [inline apps/builder/package.json:37]: We use 1.19.2, please downgrade

File: apps/builder/middleware.js [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+ * ------------------------------------------------------------
+ * ---------------------======<<<App>>>======------------------
+ * ------------------------------------------------------------
+ */
+
+/** @type {Hono<ContextEnv>} */
+const app = new Hono();
+
+// Debug route, to show current values, MUST BE ABOVE middlewares
+app.get("/rate-limit/debug", async (ctx) => {
```
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/middleware.js:229]: They are not passthrough and should not harm anything. Gonna use in next iterations

File: apps/builder/middleware.js [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+/**
+ * @typedef {object} Environment
+ * @property {string} KV_RATE_LIMIT_REST_API_TOKEN
+ * @property {string} KV_RATE_LIMIT_REST_API_URL
+ * @property {string} TRPC_SERVER_API_TOKEN
+ * @property {string} AUTH_SECRET
+ *
+ * @typedef {object} Variables
+ * @property {boolean} skipRateLimit
```
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/middleware.js:45]: Yes env

File: apps/builder/middleware.js [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+  return data.user?.id;
+};
+
+/**
+ * @param {import('hono').Context<ContextEnv>} ctx
+ * @param {keyof ReturnType<typeof getRateLimits>} ratelimitName
+ * @param {string} key
+ */
+const checkRateLimit = async (ctx, ratelimitName, key) => {
+  const rateLimits = getRateLimits(ctx);
```
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/middleware.js:175]: Yep, memoized

File: apps/builder/middleware.js [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+      redis: kv,
+      limiter: Ratelimit.slidingWindow(10, "5 s"),
+      ephemeralCache: rateLimiterCache,
+      prefix: "unauthorized",
+    });
+
+    // No ephemeralCache for the AI endpoint
+    const ai = new Ratelimit({
+      redis: kv,
+      limiter: Ratelimit.slidingWindow(20, "5 m"),
```
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/middleware.js:96]: Im thinking to add env vars for limit setup, will do now

File: apps/builder/package.json [hunk 7]
```diff
@@ -14,7 +14,10 @@
     "checks": "pnpm typecheck && pnpm test",
     "storybook:dev": "storybook dev -p 6006",
     "storybook:build": "storybook build",
-    "ci:migrate": "migrations migrate --force"
+    "ci:migrate": "migrations migrate --force",
+    "middleware:link": "pnpx vercel link",
+    "middleware:env": "pnpx vercel env pull --environment=preview ./.env.preview",
+    "middleware:dev": "DOTENV_CONFIG_PATH=.env.preview tsx watch -r dotenv/config ./middleware.js"
```
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/package.json:20]: In 99% of code edits locally we dont need this at all. Should work with vercel dev, just we don’t use it.

[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/package.json:20]: The isue with vercel dev that it's monorepo, so you need to run it from project root

[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/package.json:20]: Anyway I made support of `vercel dev`

[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/package.json:20]: I will leave node implementation as it's easier to debug.

[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/package.json:20]: I found few small bugs during making vercel support, so its fine

File: apps/builder/middleware.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/middleware.js:103]: Decided to do later

**Review by TrySound — APPROVED** [review]

**Pull request merged**